File: modules/agent/src/controllers/usage_controller.py
"""Controller layer for usage analytics routing to services."""

from typing import List

from fastapi import Request

from lib.decorators import handle_http_exceptions
from lib.role_check import user_has_role
from services import usage_service as service


@handle_http_exceptions
async def get_user_usage(request: Request, period: str = "monthly") -> dict:
    """Get aggregated usage for the current user."""
    user_id = request.state.user_id
    return await service.get_user_usage(user_id, period)


@handle_http_exceptions
async def get_tenant_usage(request: Request, period: str = "monthly") -> dict:
    """Get aggregated usage for the tenant."""
    tenant_id = request.state.tenant_id
    return await service.get_tenant_usage(tenant_id, period)


@handle_http_exceptions
async def get_usage_timeseries(
    request: Request,
    period: str = "monthly",
    scope: str = "user",
) -> List[dict]:
    """Get usage timeseries data for charts."""
    tenant_id = request.state.tenant_id
    user_id = request.state.user_id if scope == "user" else None
    return await service.get_usage_timeseries(tenant_id, period, user_id)


@handle_http_exceptions
async def get_developer_analytics(
    request: Request,
    period: str = "monthly",
    group_by: str = "node",
) -> dict:
    """Get developer analytics breakdown (requires developer role)."""
    user_id = request.state.user_id
    tenant_id = request.state.tenant_id

    has_role = await user_has_role(user_id, "developer")
    if not has_role:
        return {"error": "Insufficient permissions", "data": []}

    group_by_handlers = {
        "model": service.get_usage_by_model,
        "node": service.get_usage_by_node,
    }
    handler = group_by_handlers.get(group_by, service.get_usage_by_node)
    data = await handler(tenant_id, period)

    return {"group_by": group_by, "data": data}


@handle_http_exceptions
async def check_developer_access(request: Request) -> dict:
    """Check if user has developer role."""
    user_id = request.state.user_id
    has_role = await user_has_role(user_id, "developer")
    return {"has_developer_access": has_role}

File: modules/agent/internal/controllers/usage_controller.go
package controllers

import (
	"net/http"

	"agent/internal/middleware"
	"agent/internal/services"
)

// UsageController handles usage analytics HTTP requests
type UsageController struct {
	usageService *services.UsageService
}

// NewUsageController creates a new usage controller
func NewUsageController(usageService *services.UsageService) *UsageController {
	return &UsageController{usageService: usageService}
}

// GetUserUsage handles GET /usage/user
func (c *UsageController) GetUserUsage(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "user not found")
		return
	}

	period := r.URL.Query().Get("period")
	if period == "" {
		period = "monthly"
	}

	result, err := c.usageService.GetUserUsage(userID, period)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// GetTenantUsage handles GET /usage/tenant
func (c *UsageController) GetTenantUsage(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant not found")
		return
	}

	period := r.URL.Query().Get("period")
	if period == "" {
		period = "monthly"
	}

	result, err := c.usageService.GetTenantUsage(tenantID, period)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// GetUsageTimeseries handles GET /usage/timeseries
func (c *UsageController) GetUsageTimeseries(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant not found")
		return
	}

	period := r.URL.Query().Get("period")
	if period == "" {
		period = "monthly"
	}

	scope := r.URL.Query().Get("scope")
	var userID *int64
	uid, ok := middleware.GetUserID(r.Context())
	if (scope == "user" || scope == "") && ok {
		userID = &uid
	}

	result, err := c.usageService.GetUsageTimeseries(tenantID, period, userID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// GetDeveloperAnalytics handles GET /usage/analytics
func (c *UsageController) GetDeveloperAnalytics(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "user not found")
		return
	}

	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant not found")
		return
	}

	period := r.URL.Query().Get("period")
	if period == "" {
		period = "monthly"
	}

	groupBy := r.URL.Query().Get("group_by")
	if groupBy == "" {
		groupBy = "node"
	}

	result, err := c.usageService.GetDeveloperAnalytics(userID, tenantID, period, groupBy)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// CheckDeveloperAccess handles GET /usage/developer-access
func (c *UsageController) CheckDeveloperAccess(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "user not found")
		return
	}

	result, err := c.usageService.CheckDeveloperAccess(userID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}
